promql/engine.go

```diff
@@ -576,26 +576,37 @@ func (ng *Engine) execEvalStmt(ctx context.Context, query *query, s *parser.Eval
 	return mat, warnings, nil
 }
 
-// subqueryOffsetRange returns the sum of offsets and ranges of all subqueries in the path.
-func (ng *Engine) subqueryOffsetRange(path []parser.Node) (time.Duration, time.Duration) {
+// subqueryOffsetRangeTimestamp returns the sum of offsets and ranges of all subqueries in the path.
+// It also returns the min and max timestamp seen associated with the subquery.
+func (ng *Engine) subqueryOffsetRangeTimestamp(path []parser.Node) (time.Duration, time.Duration, int64, int64) {
 	var (
-		subqOffset time.Duration
-		subqRange  time.Duration
+		subqOffset, subqRange      time.Duration
+		minTimestamp, maxTimestamp int64 = math.MaxInt64, math.MinInt64
 	)
 	for _, node := range path {
 		switch n := node.(type) {
 		case *parser.SubqueryExpr:
 			subqOffset += n.Offset
 			subqRange += n.Range
+			if n.Timestamp > 0 {
+				ts := timestamp.FromFloatSeconds(n.Timestamp)
+				if ts < minTimestamp {
+					minTimestamp = ts
+				}
+				if ts > maxTimestamp {
+					maxTimestamp = ts
+				}
+			}
 		}
 	}
-	return subqOffset, subqRange
+	return subqOffset, subqRange, minTimestamp, maxTimestamp
 }
 
 func (ng *Engine) findMinTime(s *parser.EvalStmt) time.Time {
```

Comment:
All these changes need unittests, as remote read sees them.

Max time also now needs one of these functions, as it can be in the future.